File: src/claude_usage_monitor/firefox_cookies.py
"""Read claude.ai session cookies directly from Firefox's cookie store.

Firefox stores cookies unencrypted in an SQLite database at:
  %APPDATA%\\Mozilla\\Firefox\\Profiles\\<profile>\\cookies.sqlite

The database is opened read-only via SQLite URI (mode=ro). Firefox's WAL
journal mode allows concurrent reads without needing a file copy.
"""

from __future__ import annotations

import configparser
import os
import re
import sqlite3
import time
from pathlib import Path

_UUID_RE = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")


def _appdata_roaming() -> Path:
    # Respect %APPDATA% — profiles redirected via group policy don't live
    # under Path.home().
    appdata = os.environ.get("APPDATA")
    return Path(appdata) if appdata else Path.home() / "AppData" / "Roaming"


_FIREFOX_APPDATA = _appdata_roaming() / "Mozilla" / "Firefox"
_CLAUDE_HOST = "claude.ai"


class FirefoxCookieError(Exception):
    """Raised when cookies cannot be read from Firefox's profile."""


def _find_default_profile() -> Path:
    """Locate the default Firefox profile directory on Windows."""
    profiles_ini = _FIREFOX_APPDATA / "profiles.ini"
    if not profiles_ini.exists():
        raise FirefoxCookieError(
            f"Firefox profiles.ini not found at {profiles_ini}. "
            "Is Firefox installed?"
        )

    cfg = configparser.ConfigParser()
    cfg.read(profiles_ini, encoding="utf-8")

    # Prefer the profile named in an [Install…] section (the one the default
    # Firefox install actually uses), then the first [Profile*] section with
    # Default=1, then any existing profile.
    for section in cfg.sections():
        if not section.startswith("Install"):
            continue
        install_default = cfg[section].get("Default")
        if not install_default:
            continue
        candidate = _FIREFOX_APPDATA / install_default
        if candidate.is_dir():
            return candidate

    # Fallback: scan [Profile*] sections, honouring the Default=1 flag.
    fallback: Path | None = None
    for section in cfg.sections():
        if not section.startswith("Profile"):
            continue
        profile_cfg = cfg[section]
        relative = profile_cfg.get("IsRelative", "1") == "1"
        path_str = profile_cfg.get("Path", "")
        profile_path = (
            _FIREFOX_APPDATA / path_str if relative else Path(path_str)
        )
        if not profile_path.is_dir():
            continue
        if profile_cfg.get("Default") == "1":
            return profile_path
        if fallback is None:
            fallback = profile_path

    if fallback is not None:
        return fallback

    raise FirefoxCookieError("No usable Firefox profile directory found.")


def _query_cookies(db_path: Path, host: str) -> dict[str, str]:
    """Read all non-expired cookies for *host* directly without copying.

    Matches the exact host and any subdomain (host-only `claude.ai` and
    domain cookies stored as `.claude.ai` / `.sub.claude.ai`). An earlier
    `LIKE '%host%'` pattern would have matched unrelated domains like
    `notclaude.ai` or `evil-claude.ai.example.com`.
    """
    now_seconds = int(time.time())
    params = (host, f".{host}", f"%.{host}", now_seconds)
    try:
        conn = sqlite3.connect(f"file:{db_path}?mode=ro", uri=True)
        try:
            # Firefox containers store separate rows per container (column
            # originAttributes). Sort so default-container rows ('') come last
            # and win the dict merge — otherwise a container session could
            # shadow the regular login.
            try:
                cur = conn.execute(
                    """
                    SELECT name, value, originAttributes
                    FROM   moz_cookies
                    WHERE  (host = ? OR host = ? OR host LIKE ?)
                      AND  expiry > ?
                    """,
                    params,
                )
                rows = sorted(cur.fetchall(), key=lambda r: r[2] == "")
            except sqlite3.OperationalError:
                # Very old schema without originAttributes.
                cur = conn.execute(
                    """
                    SELECT name, value
                    FROM   moz_cookies
                    WHERE  (host = ? OR host = ? OR host LIKE ?)
                      AND  expiry > ?
                    """,
                    params,
                )
                rows = cur.fetchall()
            return {row[0]: row[1] for row in rows}
        finally:
            conn.close()
    except sqlite3.Error as exc:
        raise FirefoxCookieError(
            f"Could not read Firefox cookie database: {exc}"
        ) from exc


def get_claude_cookies(profile_dir: Path | None = None) -> dict[str, str]:
    """Return current claude.ai cookies from Firefox.

    Args:
        profile_dir: Override the auto-detected Firefox profile directory.

    Returns:
        Dict of cookie name → value for all non-expired claude.ai cookies.

    Raises:
        FirefoxCookieError: If the profile or cookie DB cannot be found/read.
    """
    resolved = profile_dir or _find_default_profile()
    db_path = resolved / "cookies.sqlite"
    if not db_path.exists():
        raise FirefoxCookieError(
            f"cookies.sqlite not found in Firefox profile: {resolved}\n"
            "Make sure Firefox has been launched at least once."
        )

    cookies = _query_cookies(db_path, _CLAUDE_HOST)
    if not cookies:
        raise FirefoxCookieError(
            "No claude.ai cookies found in Firefox. "
            "Please log in to claude.ai in Firefox first."
        )
    return cookies


class CookieError(Exception):
    """Browser-agnostic cookie error used for org-ID failures."""


def extract_org_id(cookies: dict[str, str]) -> str:
    """Pull the organisation UUID from the lastActiveOrg cookie.

    The value is validated as a canonical UUID before being returned so it
    cannot smuggle extra path segments into the API URL.
    """
    org_id = cookies.get("lastActiveOrg", "").strip('"').lower()
    if not org_id:
        raise CookieError(
            "lastActiveOrg cookie not found. "
            "Visit claude.ai in your browser to set an active organisation."
        )
    if not _UUID_RE.match(org_id):
        raise CookieError(
            "lastActiveOrg cookie is not a valid UUID. "
            "Open claude.ai in Firefox to refresh the session."
        )
    return org_id


def build_cookie_header(cookies: dict[str, str]) -> str:
    """Serialise the cookie dict to an HTTP Cookie header string."""
    return "; ".join(f"{k}={v}" for k, v in cookies.items())

File: csharp/ClaudeUsageTracker/FirefoxCookies.cs
// Read claude.ai session cookies directly from Firefox's cookie store.
//
// Firefox stores cookies unencrypted in an SQLite database at:
//   %APPDATA%\Mozilla\Firefox\Profiles\<profile>\cookies.sqlite
//
// The database is opened read-only. Firefox's WAL journal mode allows
// concurrent reads without needing a file copy.

using System.IO;
using System.Text.RegularExpressions;
using Microsoft.Data.Sqlite;

namespace ClaudeUsageTracker;

/// <summary>Raised when cookies cannot be read from Firefox's profile.</summary>
public class FirefoxCookieException : Exception
{
    public FirefoxCookieException(string message) : base(message) { }
    public FirefoxCookieException(string message, Exception inner) : base(message, inner) { }
}

/// <summary>Browser-agnostic cookie error used for org-ID failures.</summary>
public class CookieException : Exception
{
    public CookieException(string message) : base(message) { }
}

public static partial class FirefoxCookies
{
    private const string ClaudeHost = "claude.ai";

    [GeneratedRegex("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")]
    private static partial Regex UuidRegex();

    private static string FirefoxAppData
    {
        get
        {
            // Respect %APPDATA% — profiles redirected via group policy don't
            // live under the user-profile folder.
            var appdata = Environment.GetEnvironmentVariable("APPDATA");
            var baseDir = !string.IsNullOrEmpty(appdata)
                ? appdata
                : Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "AppData", "Roaming");
            return Path.Combine(baseDir, "Mozilla", "Firefox");
        }
    }

    /// <summary>Locate the default Firefox profile directory on Windows.
    /// Prefers the profile named in an [Install…] section, then the first
    /// [Profile*] section with Default=1, then any existing profile.</summary>
    internal static string FindDefaultProfile(string? firefoxAppData = null)
    {
        var appData = firefoxAppData ?? FirefoxAppData;
        var profilesIni = Path.Combine(appData, "profiles.ini");
        if (!File.Exists(profilesIni))
        {
            throw new FirefoxCookieException(
                $"Firefox profiles.ini not found at {profilesIni}. Is Firefox installed?");
        }

        var sections = ParseIni(profilesIni);

        foreach (var (name, values) in sections)
        {
            if (!name.StartsWith("Install", StringComparison.Ordinal))
                continue;
            if (!values.TryGetValue("Default", out var installDefault) || string.IsNullOrEmpty(installDefault))
                continue;
            var candidate = Path.Combine(appData, installDefault);
            if (Directory.Exists(candidate))
                return candidate;
        }

        // Fallback: scan [Profile*] sections, honouring the Default=1 flag.
        string? fallback = null;
        foreach (var (name, values) in sections)
        {
            if (!name.StartsWith("Profile", StringComparison.Ordinal))
                continue;
            var relative = values.GetValueOrDefault("IsRelative", "1") == "1";
            var pathStr = values.GetValueOrDefault("Path", "");
            var profilePath = relative ? Path.Combine(appData, pathStr) : pathStr;
            if (!Directory.Exists(profilePath))
                continue;
            if (values.GetValueOrDefault("Default") == "1")
                return profilePath;
            fallback ??= profilePath;
        }

        if (fallback is not null)
            return fallback;

        throw new FirefoxCookieException("No usable Firefox profile directory found.");
    }

    /// <summary>Minimal INI parser. Values are literal (no interpolation) —
    /// profiles.ini paths may contain '%' characters.</summary>
    private static List<(string Section, Dictionary<string, string> Values)> ParseIni(string path)
    {
        var sections = new List<(string, Dictionary<string, string>)>();
        Dictionary<string, string>? current = null;
        foreach (var rawLine in File.ReadAllLines(path))
        {
            var line = rawLine.Trim();
            if (line.Length == 0 || line.StartsWith(';') || line.StartsWith('#'))
                continue;
            if (line.StartsWith('[') && line.EndsWith(']'))
            {
                current = [];
                sections.Add((line[1..^1], current));
                continue;
            }
            var eq = line.IndexOf('=');
            if (eq > 0 && current is not null)
                current[line[..eq].Trim()] = line[(eq + 1)..].Trim();
        }
        return sections;
    }

    /// <summary>Read all non-expired cookies for <paramref name="host"/>
    /// directly, without copying the database.
    /// Matches the exact host and any subdomain (host-only `claude.ai` and
    /// domain cookies stored as `.claude.ai` / `.sub.claude.ai`).</summary>
    internal static Dictionary<string, string> QueryCookies(string dbPath, string host)
    {
        var nowSeconds = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
        try
        {
            using var conn = new SqliteConnection(new SqliteConnectionStringBuilder
            {
                DataSource = dbPath,
                Mode = SqliteOpenMode.ReadOnly,
                Pooling = false, // don't hold a handle on Firefox's DB after the poll
            }.ToString());
            conn.Open();

            List<(string Name, string Value, string OriginAttributes)> rows;
            try
            {
                rows = RunQuery(conn, host, nowSeconds, withOriginAttributes: true);
                // Firefox containers store separate rows per container (column
                // originAttributes). Sort so default-container rows ('') come
                // last and win the dict merge — otherwise a container session
                // could shadow the regular login.
                rows = [.. rows.OrderBy(r => r.OriginAttributes == "" ? 1 : 0)];
            }
            catch (SqliteException)
            {
                // Very old schema without originAttributes.
                rows = RunQuery(conn, host, nowSeconds, withOriginAttributes: false);
            }

            var cookies = new Dictionary<string, string>();
            foreach (var row in rows)
                cookies[row.Name] = row.Value;
            return cookies;
        }
        catch (SqliteException exc)
        {
            throw new FirefoxCookieException($"Could not read Firefox cookie database: {exc.Message}", exc);
        }
    }

    private static List<(string, string, string)> RunQuery(
        SqliteConnection conn, string host, long nowSeconds, bool withOriginAttributes)
    {
        var columns = withOriginAttributes ? "name, value, originAttributes" : "name, value";
        using var cmd = conn.CreateCommand();
        cmd.CommandText = $"""
            SELECT {columns}
            FROM   moz_cookies
            WHERE  (host = $host OR host = $dotHost OR host LIKE $subHost)
              AND  expiry > $now
            """;
        cmd.Parameters.AddWithValue("$host", host);
        cmd.Parameters.AddWithValue("$dotHost", $".{host}");
        cmd.Parameters.AddWithValue("$subHost", $"%.{host}");
        cmd.Parameters.AddWithValue("$now", nowSeconds);

        var rows = new List<(string, string, string)>();
        using var reader = cmd.ExecuteReader();
        while (reader.Read())
        {
            rows.Add((
                reader.GetString(0),
                reader.GetString(1),
                withOriginAttributes && !reader.IsDBNull(2) ? reader.GetString(2) : ""));
        }
        return rows;
    }

    /// <summary>Return current claude.ai cookies from Firefox.</summary>
    /// <exception cref="FirefoxCookieException">Profile or cookie DB missing/unreadable.</exception>
    public static Dictionary<string, string> GetClaudeCookies(string? profileDir = null)
    {
        var resolved = profileDir ?? FindDefaultProfile();
        var dbPath = Path.Combine(resolved, "cookies.sqlite");
        if (!File.Exists(dbPath))
        {
            throw new FirefoxCookieException(
                $"cookies.sqlite not found in Firefox profile: {resolved}\n" +
                "Make sure Firefox has been launched at least once.");
        }

        var cookies = QueryCookies(dbPath, ClaudeHost);
        if (cookies.Count == 0)
        {
            throw new FirefoxCookieException(
                "No claude.ai cookies found in Firefox. " +
                "Please log in to claude.ai in Firefox first.");
        }
        return cookies;
    }

    /// <summary>Pull the organisation UUID from the lastActiveOrg cookie.
    /// The value is validated as a canonical UUID before being returned so it
    /// cannot smuggle extra path segments into the API URL.</summary>
    public static string ExtractOrgId(IReadOnlyDictionary<string, string> cookies)
    {
        var orgId = (cookies.GetValueOrDefault("lastActiveOrg") ?? "").Trim('"').ToLowerInvariant();
        if (string.IsNullOrEmpty(orgId))
        {
            throw new CookieException(
                "lastActiveOrg cookie not found. " +
                "Visit claude.ai in your browser to set an active organisation.");
        }
        if (!UuidRegex().IsMatch(orgId))
        {
            throw new CookieException(
                "lastActiveOrg cookie is not a valid UUID. " +
                "Open claude.ai in Firefox to refresh the session.");
        }
        return orgId;
    }

    /// <summary>Serialise the cookie dict to an HTTP Cookie header string.</summary>
    public static string BuildCookieHeader(IReadOnlyDictionary<string, string> cookies) =>
        string.Join("; ", cookies.Select(kv => $"{kv.Key}={kv.Value}"));
}
